### PR TITLE
PipeNet

### PR DESCRIPTION
**Pull request by collinlunn**
WIP, but want to show changes in case I'm misusing atmos code.

Feature List:
- [x] Pipes (Straight, Bend, 4 way, etc.)
- [x] PipeNet that combines and splits gas contents as pipes are added/removed
- [x] Pumps
- [x] Vents, Scrubbers

Future:
Anchoring/Unanchoring pipes
Real pumps/vents/scrubbers
Real sprites

### CODE REVIEW

**Comment by ike709** [comment]
I'm pretty sure gasses in pipes aren't supposed to react.

File: Content.Server/GameObjects/Components/Atmos/Piping/PipeComponent.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
+﻿using Content.Server.Atmos;
+using Content.Server.GameObjects.Components.NodeContainer.NodeGroups;
+using Robust.Shared.GameObjects;
+using Robust.Shared.Serialization;
+using Robust.Shared.ViewVariables;
+
+namespace Content.Server.GameObjects.Components.Atmos
+{
+    [RegisterComponent]
+    public class PipeComponent : Component
```
Comment (Zumorica) [inline Content.Server/GameObjects/Components/Atmos/Piping/PipeComponent.cs:10]: Make this a IGasMixtureHolder, please.
AssumeAir simply merges giver mixture into the local mixture.

File: Content.Server/GameObjects/EntitySystems/PumpSystem.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+        public override void Update(float frameTime)
+        {
+            base.Update(frameTime);
+            foreach (var pump in ComponentManager.EntityQuery<BasePumpComponent>())
+            {
+                pump.Update(frameTime);
+            }
+        }
+    }
+}
```
Comment (Zumorica) [inline Content.Server/GameObjects/EntitySystems/PumpSystem.cs:17]: It'd probably be best if all of pipenet updates were on their own update stage in GridAtmosphereComponent, and have them update only the pipenets on the current grid.

File: Content.Server/GameObjects/Components/Atmos/Piping/PipeContainerComponent.cs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+        public void JoinPipeNet(IPipeNet pipeNet)
+        {
+            _pipeNet = pipeNet;
+            _needsPipeNet = false;
+        }
+
+        public void ClearPipeNet()
+        {
+            _pipeNet = NodeContainer.NodeGroups.PipeNet.NullNet;
```
Comment (RemieRichards) [inline Content.Server/GameObjects/Components/Atmos/Piping/PipeContainerComponent.cs:96]: is this different to `PipeNet.NullNet` somehow?
if not, simplify.

File: Content.Server/GameObjects/Components/NodeContainer/Nodes/PipeNode.cs [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+            West = Direction.West,
+        }
+    }
+
+    public enum PipeDirection
+    {
+        None = 0,
+
+        //Half of a pipe in a direction
+        Up = 1 << 0,
```
Comment (RemieRichards) [inline Content.Server/GameObjects/Components/NodeContainer/Nodes/PipeNode.cs:100]: I'd be careful with the terms `Up` and `Down`, given that we may eventually have Z-level pipes which would LITERALLY go `Up`/`Down`.

Just go with the standard `North`/`East`/`South`/`West` direction names

**Comment by collinlunn** [comment]
Functionality of pumps, vents, and scrubbers are going to be placeholder for now. They just transfer all gas instantly, can be replaced later.

File: Content.Server/GameObjects/Components/Atmos/GridAtmosphereComponent.cs [hunk 5]
```diff
@@ -509,6 +559,46 @@ private void ProcessSuperconductivity()
...
+
+                if (number++ < LagCheckIterations) continue;
+                number = 0;
+                // Process the rest next time.
+                if (_stopwatch.Elapsed.TotalMilliseconds >= LagCheckMaxMilliseconds)
+                    return;
+            }
+            _deviceIndex = 0;
+        }
+
```
Comment (collinlunn) [inline Content.Server/GameObjects/Components/Atmos/GridAtmosphereComponent.cs:601]: Looking for particular review of the ds & algo here

Comment (collinlunn) [reply Content.Server/GameObjects/Components/Atmos/GridAtmosphereComponent.cs:601]: Good catch, typo'd the same index var

**Pull request merged**